src/haxelib/api/ScriptRunner.hx

```diff
@@ -0,0 +1,183 @@
+package haxelib.api;
+
+import sys.FileSystem;
+
+import haxelib.Data.DependencyVersion;
+import haxelib.api.LibraryData;
+
+/** Contains data with which a library script is executed. **/
+@:structInit
+class CallData {
+	/** Directory in which the script will start. Defaults to current working directory. **/
+	public final dir = Sys.getCwd();
+	/** Array of arguments to be passed on to the library call. **/
+	public final args:Array<String> = [];
+	/**
+		Whether to pass `--haxelib-global` to the haxe compiler when running
+		(only works with haxe 4.0.0 and above).
+	**/
+	public final useGlobalRepo:Bool = false;
+}
+
+@:noDoc
+/** either the project `name` or `name:version` **/
+abstract Dependency(String) from ProjectName to String {
+	inline function new(d) this = d;
+	public static function fromNameAndVersion(name:ProjectName, version:DependencyVersion):Dependency
+		return new Dependency(switch cast(version, String) {
+			case '': name;
+			case version: '$name:$version';
+		});
+}
+
+@:noDoc
+typedef Dependencies = Array<Dependency>;
+
+@:noDoc
+/** Library data needed in order to run it **/
+typedef LibraryRunData = {
+	/** This may be an alias. **/
+	final name:ProjectName;
+	/** This is the actual name found in the `haxelib.json`. **/
+	final internalName:ProjectName;
+	final version:VersionOrDev;
+	final dependencies:Dependencies;
+	final main:Null<String>;
+	final path:String;
+}
+
+private enum RunType {
+	/** Runs compiled neko file at `path` **/
+	Neko(path:String);
+	/**  **/
+	Script(main:String, name:ProjectName, version:VersionOrDev, dependencies:Dependencies);
+}
+
+private typedef State = {
+	final dir:String;
+	final run:Null<String>;
+	final runName:Null<String>;
+}
+
+/** Exception which is thrown if running a library script returns a non-zero code. **/
+@:noDoc
+class ScriptError extends haxe.Exception {
+	/** The error code returned by the library script call. **/
+	public final code:Int;
+	public function new(name:ProjectName, code:Int) {
+		super('Script for library "$name" exited with error code: $code');
+		this.code = code;
+	}
+}
+
+/** Class containing function for running a library's script. **/
+@:noDoc
+class ScriptRunner {
+	static final HAXELIB_RUN = "HAXELIB_RUN";
+	static final HAXELIB_RUN_NAME = "HAXELIB_RUN_NAME";
+
+	/** Run `library`, with `callData`.
+
+		`compilerData` is used if it is an interpreted script.
+	 **/
+	public static function run(library:LibraryRunData, compilerData:LibraryData, callData:CallData):Void {
+		final type = getType(library);
+
+		final cmd = getCmd(type);
+		final args = generateArgs(type, callData, SemVer.ofString(compilerData.version));
+
+		final oldState = getState();
+
+		// call setup
+		setState({
+			dir: library.path,
+			run: "1",
+			runName: library.internalName
+		});
+
+		final output = Sys.command(cmd, args);
```

Comment:
Are we sure this doesn't throw? If it does, `setState` below is missed.

Reply:
Theoretically Sys.command doesn't throw. I could wrap it in try catch just in case?